### PR TITLE
Torchfx gptq pattern replacer

### PR DESCRIPTION
**Pull request by cavusmustafa**
Required for GPTQ int4 LLMs. (ChatGLM2, ChatGLM3, Llama2)
Includes the transformations below:
- GPTQ Decompression pattern replacer

image

- GPTQ Multiplication pattern replacer

image

Ticket: CVS-139181

### CODE REVIEW

**Comment by slyalin** [comment]
Do we know a reason why the patterns look significantly different from what we have on ts side if the same model is used as an input? @mvafin, do we have tests for GPTQ models enabled for ts path and do they still pass with the recent versions of everything?

We are patching the model in ts path with our implementation of MatMul with quantization, and this can be a reason that we have different patterns. But we are patching because the source model cannot be traced. @cavusmustafa, as you are not patching the model on the fx path but still have some patterns that entirely represent area of the interest in the graph, that means that the initial model can be successfully traced in fx and patching is not required for tracing. @cavusmustafa, have you tried ts path for the same model with disabled pathing inside ts decoder as we discussed some time ago? If the model doesn't need to be patched and can be traced in ts, then we can apply the same approach implemented in this PR for ts path as well and remove GPTQ patching from ts decoder. @mvafin, could you help @cavusmustafa check that?

**Comment by mvafin** [comment]
@slyalin Yes, we have GPTQ model in tests for TS path: https://github.com/openvinotoolkit/openvino/blob/a1de1908bfd5f597d48422f21796daa3bfd08120/tests/model_hub_tests/pytorch/test_hf_transformers.py#L566
It is in precommit, so it definitely works still.

File: src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/op_support.py [hunk 1]
```diff
@@ -76,9 +76,11 @@ def __init__(self, options):
             "torch.ops.aten.avg_pool2d.default": None,
             "torch.ops.aten.avg_pool3d.default": None,
             "torch.ops.aten.baddbmm.default": None,
+            "torch.ops.aten.bitwise_and.Scalar": None,
             "torch.ops.aten.bitwise_and.Tensor": None,
             "torch.ops.aten.bitwise_not.default": None,
             "torch.ops.aten.bitwise_or.Tensor": None,
+            "torch.ops.aten.bitwise_right_shift.Tensor": None,
```
Comment (mvafin) [inline src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/op_support.py:83]: `bitwise_right_shift` is only supported in GPTQ pattern, right? Should we put a condition here to check that it is used in that pattern?

File: src/frontends/pytorch/src/frontend.cpp [hunk 2]
```diff
@@ -173,6 +174,9 @@ std::shared_ptr<Model> FrontEnd::decode(const InputModel::Ptr& model) const {
 void FrontEnd::normalize(const std::shared_ptr<ov::Model>& model) const {
     ov::pass::Manager manager;
 
+    manager.register_pass<ov::frontend::pytorch::pass::GPTQDecompressionReplacer>();
+    manager.register_pass<ov::frontend::pytorch::pass::GPTQMultPatternReplacer>();
```
[the remark this answers is not in the document]
Comment (cavusmustafa) [reply src/frontends/pytorch/src/frontend.cpp:178]: It seems like if I move it after the line below, the pattern is already modified by previous transformations and the GPTQ transformation cannot capture the pattern anymore. Probably, we may have an option to update the transformation so we can move it near "U4BlockRepack". But in this case, any future changes in the previous transformations could require us to update GPTQ transformations accordingly.
https://github.com/openvinotoolkit/openvino/blob/4f5f1b1b14bf779e2988343426524c9f5c9b2e35/src/frontends/pytorch/src/frontend.cpp#L183

So, we can definitely move it a little down, just below the "AtenIndexToSelect" transformation. But I am not sure if it makes sense to move it further down due to the reason I mention above.

File: src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/op_support.py (hunk 1 above)
Comment (cavusmustafa) [reply src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/op_support.py:83]: Right, this is only supported with GPTQ pattern. I will try to implement a pass to exclude "bitwise_right_shift" if it's not part of the GPTQ pattern.

**Comment by cavusmustafa** [comment]
> Do we know a reason why the patterns look significantly different from what we have on ts side if the same model is used as an input? @mvafin, do we have tests for GPTQ models enabled for ts path and do they still pass with the recent versions of everything?
> 
> We are patching the model in ts path with our implementation of MatMul with quantization, and this can be a reason that we have different patterns. But we are patching because the source model cannot be traced. @cavusmustafa, as you are not patching the model on the fx path but still have some patterns that entirely represent area of the interest in the graph, that means that the initial model can be successfully traced in fx and patching is not required for tracing. @cavusmustafa, have you tried ts path for the same model with disabled pathing inside ts decoder as we discussed some time ago? If the model doesn't need to be patched and can be traced in ts, then we can apply the same approach implemented in this PR for ts path as well and remove GPTQ patching from ts decoder. @mvafin, could you help @cavusmustafa check that?

I didn't need to patch the model as the decompression pattern was already a part of TorchFX graph. So, the pattern shown in the first drawing above is already provided as part of the graph. For TS backend, I did a quick test by removing patching from TS decoder and comparing the performance. I did not observe any performance difference but deeper analysis should be done for sure. But should we need to clarify TS side of the issue for this PR, or can it be part of another PR if any fix is needed?

File: src/frontends/pytorch/src/frontend.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvafin) [reply src/frontends/pytorch/src/frontend.cpp:177]: Okay, that can be an effect of `ConstantFolding` transformation, let's leave it in the beginning, but please add comment why

File: src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/op_support.py (hunk 1 above)
Comment (cavusmustafa) [reply src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/op_support.py]: Implemented an additional pass before partitioning. Now, bitwise_right_shift is disabled by default. If any GPTQ pattern is found by the pass, the ops inside the pattern are individually enabled by their ids. This way, bitwise_right_shift should be enabled only if it is part of a GPTQ pattern.

**Review by suryasidd — APPROVED** [review]
All the comments have been addressed. LGTM.

**Comment by suryasidd** [comment]
build_jenkins

**Comment by suryasidd** [comment]
build_jenkins

**Comment by cavusmustafa** [comment]
Please add label for code freeze

**Comment by suryasidd** [comment]
build_jenkins

**Review by mvafin — APPROVED** [review]
@cavusmustafa Could you add test for GPTQ model? Can be a GHA test, but not in scope of this PR.

**Comment by cavusmustafa** [comment]
> @cavusmustafa Could you add test for GPTQ model? Can be a GHA test, but not in scope of this PR.

Sure, I will  add tests in a later PR

**Pull request merged**